pkg/sql/catalog/tabledesc/fix_cast_for_style_visitor.go

```diff
@@ -162,7 +161,7 @@ func ResolveCastForStyleUsingVisitor(
 ) (tree.Expr, bool, error) {
 	v := makeFixCastForStyleVisitor(ctx, semaCtx)
 
-	descBuilder := tabledesc.NewBuilder(desc)
+	descBuilder := NewBuilder(desc)
```

Comment:
please use `desc.NewBuilder()`

Reply:
I can't comment on the below lines, but, do we need to use a typed expr? Can we instead just delegate to the parser and have it use an untyped and qualified expr?

Reply:
yeah, we do need a typed expr, since the visitor needs to look at the input and output types of cast expressions to determine if they need to be rewritten

Reply:
I'm not sure what to do here. We may need to find a way to plumb a fake type resolver that somehow works. This all makes me a tad nervous.

Reply:
Getting that fake type resolver into `RunPostDeserializationChanges` sounds hard.

But what if we change it so the migration doesn't use `RunPostDeserializationChanges` and instead modifies the descriptor separately? Then it would be pretty easy to make another resolver that works just like `importTypeResolver`: https://github.com/cockroachdb/cockroach/blob/d68a48a7b9990442a9131a98845ef315569e6689/pkg/sql/importer/import_type_resolver.go#L30

Maybe this is a bad idea though - it means that the invalid cast expressions would not get rewritten during a RESTORE from an older backup.

Reply:
better idea: let's try this diff
```
diff --git a/pkg/sql/catalog/tabledesc/fix_cast_for_style_visitor.go b/pkg/sql/catalog/tabledesc/fix_cast_for_style_visitor.go
index 3df28993a4..03c62efb83 100644
--- a/pkg/sql/catalog/tabledesc/fix_cast_for_style_visitor.go
+++ b/pkg/sql/catalog/tabledesc/fix_cast_for_style_visitor.go
@@ -72,6 +72,14 @@ func (v *fixCastForStyleVisitor) VisitPost(expr tree.Expr) tree.Expr {
 	}
 
 	if expr, ok := expr.(*tree.CastExpr); ok {
+		typedExpr, err := schemaexpr.DequalifyAndTypeCheckExpr(
+			v.ctx, v.tDesc, expr, v.semaCtx,
+			tree.NewUnqualifiedTableName(tree.Name(v.tDesc.GetName())),
+		)
+		if err != nil {
+			return expr
+		}
+		expr = typedExpr.(*tree.CastExpr)
 		sd := sessiondata.SessionData{
 			SessionData: sessiondatapb.SessionData{
 				IntervalStyleEnabled: v.semaCtx.IntervalStyleEnabled,
@@ -165,11 +173,6 @@ func ResolveCastForStyleUsingVisitor(
 	tDesc := descBuilder.BuildImmutableTable()
 	v.tDesc = tDesc
 
-	typedExpr, err := schemaexpr.DequalifyAndTypeCheckExpr(ctx, tDesc, expr, semaCtx, tn)
-	if err != nil {
-		v.err = err
-		return nil, false, err
-	}
-	expr, changed := tree.WalkExpr(&v, typedExpr)
+	expr, changed := tree.WalkExpr(&v, expr)
 	return expr, changed, v.err
 }
```

It delays the type-checking part so it only happens when we need it -- i.e., when the visitor is analyzing a CastExpr.